File: modules/publix/app/session/batch/akka/actors/BatchActionHandler.java
package session.batch.akka.actors;

import java.io.IOException;

import javax.inject.Inject;
import javax.inject.Singleton;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.github.fge.jsonpatch.JsonPatch;
import com.github.fge.jsonpatch.JsonPatchException;

import daos.common.BatchDao;
import models.common.Batch;
import play.Logger;
import play.Logger.ALogger;
import play.db.jpa.JPAApi;
import play.libs.Json;
import session.Registry;
import session.batch.akka.actors.BatchDispatcherProtocol.BatchActionMsg;
import session.batch.akka.actors.BatchDispatcherProtocol.BatchActionMsg.BatchAction;
import session.batch.akka.actors.BatchDispatcherProtocol.BatchActionMsg.TellWhom;

/**
 * @author Kristian Lange (2017)
 */
@Singleton
public class BatchActionHandler {

	private static final ALogger LOGGER = Logger.of(BatchActionHandler.class);

	private final JPAApi jpa;

	private final BatchDao batchDao;

	private final BatchActionMsgBuilder batchActionMsgBuilder;

	@Inject
	public BatchActionHandler(JPAApi jpa, BatchDao batchDao,
			BatchActionMsgBuilder batchActionMsgBuilder) {
		this.jpa = jpa;
		this.batchDao = batchDao;
		this.batchActionMsgBuilder = batchActionMsgBuilder;
	}

	/**
	 * Handles batch actions originating from a client
	 */
	public BatchActionMsgBundle handleBatchActionMsg(long batchId,
			long studyResultId, Registry batchRegistry, ObjectNode jsonNode) {
		LOGGER.debug(
				".handleBatchActionMsg:"
						+ " batchId {}, studyResultId {}, jsonNode {}",
				batchId, studyResultId, Json.stringify(jsonNode));
		String action = jsonNode.get(BatchActionMsg.ACTION).asText();
		switch (BatchAction.valueOf(action)) {
		case SESSION:
			return handleBatchActionSessionPatch(batchId, studyResultId,
					batchRegistry, jsonNode);
		default:
			String errorMsg = "Unknown action " + action;
			BatchActionMsg msg = batchActionMsgBuilder.buildError(batchId,
					errorMsg, TellWhom.SENDER_ONLY);
			return BatchActionMsgBundle.build(msg);
		}
	}

	/**
	 * Persists batch session and tells everyone
	 */
	private BatchActionMsgBundle handleBatchActionSessionPatch(long batchId,
			long studyResultId, Registry batchRegistry, ObjectNode jsonNode) {
		return jpa.withTransaction(() -> {
			Long clientsVersion = Long.valueOf(jsonNode
					.get(BatchActionMsg.BATCH_SESSION_VERSION).asText());
			JsonNode batchSessionPatchNode = jsonNode
					.get(BatchActionMsg.BATCH_SESSION_PATCH);

			// Get batch from DB
			Batch batch = batchDao.findById(batchId);
			if (batch == null) {
				String errorMsg = "Couldn't find batch with ID " + batchId
						+ " in database.";
				BatchActionMsg msg = batchActionMsgBuilder.buildError(batchId,
						errorMsg, TellWhom.SENDER_ONLY);
				return BatchActionMsgBundle.build(msg);
			}

			// Apply patch
			JsonNode patchedSessionData;
			try {
				patchedSessionData = patchBatchSessionData(batch,
						batchSessionPatchNode);
			} catch (IOException | JsonPatchException e) {
				LOGGER.debug(".handleBatchActionSessionPatch: ", e);
				BatchActionMsg msg = batchActionMsgBuilder.buildSimple(batch,
						BatchAction.SESSION_FAIL, TellWhom.SENDER_ONLY);
				return BatchActionMsgBundle.build(msg);
			}

			LOGGER.debug(
					".handleBatchActionSessionPatch:"
							+ " batchId {}, clientsVersion {},"
							+ " batchSessionPatch {}, updatedSessionData {}",
					batchId, clientsVersion,
					Json.stringify(batchSessionPatchNode),
					Json.stringify(patchedSessionData));

			boolean success = checkVersionAndPersistBatchSessionData(batch,
					clientsVersion, patchedSessionData);
			if (success) {
				BatchActionMsg msg1 = batchActionMsgBuilder.buildSessionPatch(
						batch, batchSessionPatchNode,
						TellWhom.ALL_BUT_SENDER);
				BatchActionMsg msg2 = batchActionMsgBuilder.buildSimple(batch,
						BatchAction.SESSION_ACK, TellWhom.SENDER_ONLY);
				return BatchActionMsgBundle.build(msg1, msg2);
			} else {
				BatchActionMsg msg = batchActionMsgBuilder.buildSimple(batch,
						BatchAction.SESSION_FAIL, TellWhom.SENDER_ONLY);
				return BatchActionMsgBundle.build(msg);
			}
		});
	}

	private JsonNode patchBatchSessionData(Batch batch,
			JsonNode batchSessionPatchNode)
			throws IOException, JsonPatchException {
		JsonPatch batchSessionPatch = JsonPatch.fromJson(batchSessionPatchNode);
		JsonNode currentBatchSessionData = Json.mapper()
				.readTree(batch.getBatchSessionData());
		return batchSessionPatch.apply(currentBatchSessionData);
	}

	/**
	 * Persists the given sessionData in the Batch and does the versioning: and
	 * increases the batchSessionVersion by 1 - but only if the stored version
	 * is equal to the received one. Returns true if this was successful -
	 * otherwise false.
	 */
	private boolean checkVersionAndPersistBatchSessionData(Batch batch,
			Long version, JsonNode sessionData) {
		if (batch != null && version != null && sessionData != null
				&& batch.getBatchSessionVersion().equals(version)) {
			batch.setBatchSessionData(sessionData.toString());
			long newVersion = batch.getBatchSessionVersion() + 1l;
			batch.setBatchSessionVersion(newVersion);
			updateBatch(batch);
			return true;
		}
		return false;
	}

	/**
	 * Persists the changes in the batch.
	 */
	private void updateBatch(Batch batch) {
		jpa.withTransaction(() -> {
			batchDao.update(batch);
		});
	}

}

File: modules/session/app/batch/BatchActionHandler.scala
package batch

import batch.BatchDispatcher.{BatchAction, BatchActionJsonKey, BatchMsg, TellWhom}
import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
import com.flipkart.zjsonpatch.JsonPatch
import com.google.common.base.Strings
import daos.common.BatchDao
import models.common.Batch
import play.api.Logger
import play.api.libs.json.Reads._
import play.api.libs.json.{JsObject, JsValue, Json}

import javax.inject.{Inject, Singleton}
import scala.annotation.tailrec

/**
 * Handles batch action messages received by a BatchDispatcher from a client via a batch channel.
 */
@Singleton
class BatchActionHandler @Inject()(batchDao: BatchDao,
                                   msgBuilder: BatchActionMsgBuilder) {

  private val logger: Logger = Logger(this.getClass)

  private val maxUpdateAttempts = 5

  private val objectMapper = new ObjectMapper()

  /**
   * Handles batch action messages originating from a client: Gets a BatchMsg that contains a field
   * 'action' in their JSON. The only action handled here is the patch for the batch session.
   * The function returns BatchMsges that will be sent out to the batch members.
   */
  def handleActionMsg(actionMsg: BatchMsg, batchId: Long): List[BatchMsg] = {
    val actionValue = (actionMsg.json \ BatchActionJsonKey.Action.toString).as[String]
    val actionOpt = BatchAction.values.find(_.toString == actionValue)
    actionOpt match {
      case Some(BatchAction.Session) => handlePatch(actionMsg.json, batchId)
      case _ =>
        List(msgBuilder.buildError(s"Unknown action $actionValue", TellWhom.SenderOnly))
    }
  }

  private def handlePatch(json: JsObject, batchId: Long): List[BatchMsg] = {
    val sessionActionId = (json \ BatchActionJsonKey.SessionActionId.toString).as[Long]
    val clientsVersion = (json \ BatchActionJsonKey.SessionVersion.toString).as[Long]
    val versioning = (json \ BatchActionJsonKey.SessionVersioning.toString).as[Boolean]
    val patches = (json \ BatchActionJsonKey.SessionPatches.toString).get

    tryUpdate(
      batchId = batchId,
      sessionActionId = sessionActionId,
      clientsVersion = clientsVersion,
      versioning = versioning,
      patches = patches,
      attempt = 1
    )
  }

  /**
   * Attempts to update the batch session data for a specified batch. The update process verifies version compatibility
   * and applies the patches if possible. In case of update conflicts, retry logic is implemented based on the number of
   * attempts.
   *
   * @return a list of `BatchMsg` instances representing the results of the update process
   */
  @tailrec
  private def tryUpdate(batchId: Long, sessionActionId: Long, clientsVersion: Long, versioning: Boolean,
                        patches: JsValue, attempt: Int): List[BatchMsg] = {
    val batch = batchDao.findById(batchId)
    if (batch == null) {
      return List(msgBuilder.buildError(s"Couldn't find batch with ID $batchId in database.", TellWhom.SenderOnly))
    }

    val currentVersion = batch.getBatchSessionVersion

    // Reject a stale version before attempting to apply its patch.
    if (versioning && clientsVersion != currentVersion) {
      val errorMsg = s"Version mismatch (client version: $clientsVersion, current: $currentVersion)."
      return List(msgBuilder.buildSimple(batch, BatchAction.SessionFail, sessionActionId, Some(errorMsg), TellWhom.SenderOnly))
    }

    val patchedSessionData =
      try {
        patchSessionData(patches, batch)
      } catch {
        case e: Exception =>
          logger.debug(s".tryUpdate: batchId $batchId, patches ${Json.stringify(patches)}, " +
            s"${e.getClass.getName}: ${e.getMessage}")
          val errorMsg = s"Failed to apply patch: ${e.getMessage}"
          return List(msgBuilder.buildSimple(batch, BatchAction.SessionFail, sessionActionId, Some(errorMsg), TellWhom.SenderOnly))
      }

    logger.debug(s".tryUpdate: batchId $batchId, clientsVersion $clientsVersion, versioning $versioning, " +
      s"batchSessionPatch ${Json.stringify(patches)}, updatedSessionData ${Json.stringify(patchedSessionData)}")

    val newVersion = batchDao.updateBatchSession(batchId, currentVersion, Json.stringify(patchedSessionData))

    if (newVersion != null) {
      // `batch` is detached and only used to build the outgoing messages. Make it reflect what was successfully committed.
      batch.setBatchSessionData(Json.stringify(patchedSessionData))
      batch.setBatchSessionVersion(newVersion)

      val updateMsg = msgBuilder.buildSessionPatch(batch, patches, TellWhom.All)
      val acknowledgement = msgBuilder.buildSimple(batch, BatchAction.SessionAck, sessionActionId, None, TellWhom.SenderOnly)
      List(updateMsg, acknowledgement)

    } else if (!versioning && attempt < maxUpdateAttempts) {
      // Another transaction won. Reload the latest state, reapply the patch to that state, and try another compare-and-set.
      tryUpdate(batchId, sessionActionId, clientsVersion, versioning, patches, attempt + 1)

    } else {
      // Either versioning was enabled, or all retry attempts were exhausted.
      val currentBatch = batchDao.findById(batchId)
      val actualVersion = Option(currentBatch).map(_.getBatchSessionVersion).getOrElse(currentVersion)

      val errorMsg =
        if (versioning)
          s"Concurrent update conflict (client version: $clientsVersion, current: $actualVersion)."
        else
          s"Couldn't update batch session after $maxUpdateAttempts attempts because of concurrent updates."
      val messageBatch = Option(currentBatch).getOrElse(batch)
      List(msgBuilder.buildSimple(messageBatch, BatchAction.SessionFail, sessionActionId, Some(errorMsg), TellWhom.SenderOnly))
    }
  }

  private def patchSessionData(patches: JsValue, batch: Batch): JsValue = {
    val currentSessionData =
      if (!Strings.isNullOrEmpty(batch.getBatchSessionData))
        Json.parse(batch.getBatchSessionData)
      else
        Json.obj()

    val sourceNode: JsonNode = objectMapper.readTree(Json.stringify(currentSessionData))
    val patchNode: JsonNode = objectMapper.readTree(Json.stringify(patches))

    val resultNode = JsonPatch.apply(patchNode, sourceNode)

    Json.parse(resultNode.toString)
  }

}
